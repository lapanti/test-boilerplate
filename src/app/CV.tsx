import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'

import Header from '../components/Header'
import Main from '../components/Main'
import type { Tech } from '../constants/jobs'
import {
    jobsArraySelector,
    statusSelector,
    techsSelector,
    totalExperienceSelector,
} from '../store/ducks/experience/selectors'
import experienceSlice from '../store/ducks/experience/slice'
import Job from './cv/Job'
import Technology from './cv/Technology'

const TechnologyList = styled.ul`
    list-style-type: none;
    padding: 0;
`

const CV: React.FunctionComponent = () => {
    const dispatch = useDispatch()
    const status = useSelector(statusSelector)
    const jobs = useSelector(jobsArraySelector)
    const techExperiences = useSelector(techsSelector)
    const totalExperience = useSelector(totalExperienceSelector)
    useEffect(() => {
        if (status === '') dispatch(experienceSlice.actions.initializeStart())
    }, [status, dispatch])
    return (
        <>
            <Header subtitle="what I have done" title="CV" />
            <Main>
                <p>
                    I started out as most people, doing websites in my early teens (I&apos;ll admit, my first ones were
                    with FrontPage) for my bands etc. I once even borrowed (and did all the exercises) from the library
                    a book that detailed the then popular HTML, PHP, MySQL stack, so a lot of the websites I did back
                    then used PHP on the level of <code>include</code>. Luckily I outgrew that and started dabbling in
                    jQuery and finally got a formal education as well.
                </p>
                <p>
                    That all lead into a (tech) master&apos;s degree in Information Networks from Aalto University,
                    where I majored in ICT in business (basically all the software courses I could find). Apart from
                    that I have accumulated some professional experience in the field
                    {totalExperience
                        ? ` (${totalExperience.years} years and ${totalExperience.months} months, to be exactish)`
                        : ''}
                    .
                </p>
                {status === 'loading' && <p>Loading...</p>}
                {status === 'done' && (
                    <>
                        <h2>Technologies</h2>
                        <TechnologyList>
                            {Object.entries(techExperiences)
                                .sort()
                                .map(([key, duration], i) => (
                                    <Technology
                                        key={i}
                                        duration={duration}
                                        tech={key as Tech}
                                        totalExperience={totalExperience}
                                    />
                                ))}
                        </TechnologyList>

                        <h2>Jobs</h2>
                        {jobs.map((job, i) => (
                            <Job key={i} job={job} />
                        ))}

                        <h2>Other technologies</h2>
                        <p>
                            I have also used or dabbled with some other technologies, listed below, but I do not
                            consider myself fluent (or willing) to work with them professionally:{' '}
                        </p>
                        <TechnologyList>
                            <li>Docker</li>
                            <li>Java</li>
                            <li>Python</li>
                            <li>PHP</li>
                            <li>MySQL</li>
                            <li>PostgreSQL</li>
                            <li>React Native</li>
                            <li>Android</li>
                            <li>And the list goes on and on</li>
                        </TechnologyList>
                    </>
                )}
            </Main>
        </>
    )
}

export default CV
