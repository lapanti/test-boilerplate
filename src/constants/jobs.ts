export const techs = {
    TypeScript: 'TypeScript',
    JavaScript: 'JavaScript',
    'Node.js': 'Node.js',
    HTML: 'HTML',
    CSS: 'CSS',
    React: 'React',
    redux: 'redux',
    'redux-observable': 'redux-observable',
    'redux-saga': 'redux-saga',
    Scala: 'Scala',
} as const

export type Tech = typeof techs[keyof typeof techs]

export type JobKey =
    | 'nokiaSolutionsAndNetworks'
    | 'pulmatonSolutionsOy'
    | 'futurice'
    | 'zalando'
    | 'goforeSolinorOy'
    | 'verkkokauppaComOyj'

export interface Job {
    id: JobKey
    companyName: string
    title: string
    description: string
    techs: Tech[]
    startDate: number
    endDate?: number
}

export const jobsArray: Job[] = [
    {
        id: 'nokiaSolutionsAndNetworks',
        companyName: 'Nokia Solutions and Networks',
        title: 'Seasonal trainee',
        description: 'Programming in Research- & Development environment with Python.',
        techs: [techs.JavaScript, techs.HTML, techs.CSS],
        startDate: 1367355600000, // May 2013
        endDate: 1377982800000, // September 2013
    },
    {
        id: 'pulmatonSolutionsOy',
        companyName: 'Pulmaton Solutions Oy',
        title: 'Software Developer',
        description: 'Development of a cloud-hosted reporting tool with Java, HTML, JavaScript and CSS.',
        techs: [techs.JavaScript, techs.HTML, techs.CSS],
        startDate: 1398891600000, // May 2014
        endDate: 1406840400000, // August 2014
    },
    {
        id: 'futurice',
        companyName: 'Futurice',
        title: 'Software Developer',
        description:
            'Development in agile software projects, ranging from Android and frontend to backend. Mainly focusing on backend development with Scala and customer communication. Also worked as a tech lead in a medium-sized challenging customer project.',
        techs: [
            techs.JavaScript,
            techs['Node.js'],
            techs.HTML,
            techs.CSS,
            techs.React,
            techs.redux,
            techs['redux-observable'],
            techs.Scala,
        ],
        startDate: 1430427600000, // May 2015
        endDate: 1485900000000, // February 2017
    },
    {
        id: 'zalando',
        companyName: 'Zalando SE',
        title: 'SW Engineer',
        description:
            'Working mainly with building multiple internal tools for fashion domain experts using React, TypeScript etc. in an agile fashion, including a tool with 2-dimensional graph visualisation using D3 and PIXI.js. I also interviewed frontend/full-stack developers.',
        techs: [
            techs.TypeScript,
            techs.JavaScript,
            techs['Node.js'],
            techs.HTML,
            techs.CSS,
            techs.React,
            techs.redux,
            techs['redux-observable'],
        ],
        startDate: 1488319200000, // March 2017
        endDate: 1514757600000, // January 2018
    },
    {
        id: 'goforeSolinorOy',
        companyName: 'Gofore / Solinor Oy',
        title: 'Software Developer',
        description:
            'I worked at Solinor for a bit over one year, after which Gofore bought the company. During my whole tenure I worked at two large PaaS projects, using TypeScript, JavaScript, Node.js, React, redux, etc. At Solinor I interviewed candidates in all rounds, at Gofore I only did technical interviews.',
        techs: [
            techs.TypeScript,
            techs.JavaScript,
            techs['Node.js'],
            techs.HTML,
            techs.CSS,
            techs.React,
            techs.redux,
            techs['redux-saga'],
        ],
        startDate: 1514757600000, // January 2018
        endDate: 1569877200000, // October 2019
    },
    {
        id: 'verkkokauppaComOyj',
        companyName: 'Verkkokauppa.com Oyj',
        title: 'Software Developer',
        description:
            'Maintaining and developing new features for the Verkkokauppa.com webstore, using TypeScript, JavaScript, React, redux, HTML, CSS and more.',
        techs: [
            techs.TypeScript,
            techs.JavaScript,
            techs['Node.js'],
            techs.HTML,
            techs.CSS,
            techs.React,
            techs.redux,
            techs['redux-saga'],
        ],
        startDate: 1569877200000, // October 2019
    },
]
