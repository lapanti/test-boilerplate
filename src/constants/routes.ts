const whoami = { exact: true, path: '/', title: 'whoami' } as const
const cv = { exact: true, path: '/cv', title: 'CV' } as const
const setup = { exact: true, path: '/setup', title: 'setup' } as const

export const routesObj = {
    cv,
    setup,
    whoami,
} as const

export const routesArray = [whoami, cv, setup] as const
