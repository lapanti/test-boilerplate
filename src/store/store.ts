import { combineReducers, configureStore } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
import { all, call } from 'redux-saga/effects'

import experienceSagas from './ducks/experience/sagas'
import experienceSlice, { initialState as experience } from './ducks/experience/slice'
import themeSlice, { initialState as theme } from './ducks/theme/slice'

const reducer = combineReducers({ experience: experienceSlice.reducer, theme: themeSlice.reducer })

export type State = ReturnType<typeof reducer>

const sagaMiddleware = createSagaMiddleware()

const rootSaga = function* () {
    yield all([experienceSagas].map((saga) => call(saga)))
}

const store = configureStore({
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({ thunk: false }).concat(sagaMiddleware),
    reducer,
})

sagaMiddleware.run(rootSaga)

type PartialState = Partial<{ experience: Partial<typeof experience>; theme: Partial<typeof theme> }>

export const mockState = (state: PartialState): State => ({
    experience: { ...experience, ...state?.experience },
    theme: { ...theme, ...state?.theme },
})

export const mockStore = (state: PartialState = {}): typeof store =>
    configureStore({ preloadedState: mockState(state), reducer })

export default store
