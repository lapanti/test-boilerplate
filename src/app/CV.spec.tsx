import { render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'

import { jobsArray, Tech, techs } from '../constants/jobs'
import type { JobWithDuration } from '../store/ducks/experience/slice'
import experienceSlice from '../store/ducks/experience/slice'
import { mockState, mockStore } from '../store/store'
import CV from './CV'

jest.mock('../lib/preference', () => ({ prefersDarkMode: jest.fn().mockReturnValue(true) }))
jest.mock('./cv/Job', () => ({ job }: { job: JobWithDuration }) => <span>{job.id}</span>)
jest.mock(
    './cv/Technology',
    () => ({ tech, duration, totalExperience }: { tech: Tech; duration?: Duration; totalExperience?: Duration }) => (
        <span>
            {tech}
            {duration ? ` ${duration.years} ${duration.months}` : ''}
            {totalExperience ? ` ${totalExperience.years} ${totalExperience.months}` : ''}
        </span>
    )
)

describe('<CV />', () => {
    it('should match snapshot when status === ""', () => {
        const startSpy = jest.spyOn(experienceSlice.actions, 'initializeStart')
        const { container } = render(
            <Provider store={mockStore(mockState({ experience: { status: '' } }))}>
                <CV />
            </Provider>
        )

        expect(screen.getByRole('heading', { name: /CV/i })).toBeInTheDocument()
        expect(screen.getByText('what I have done')).toBeInTheDocument()

        expect(startSpy).toHaveBeenCalledTimes(1)

        expect(container).toMatchSnapshot()
    })

    it('should match snapshot when status === "loading"', () => {
        const startSpy = jest.spyOn(experienceSlice.actions, 'initializeStart')
        const { container } = render(
            <Provider store={mockStore(mockState({ experience: { status: 'loading' } }))}>
                <CV />
            </Provider>
        )

        expect(screen.getByRole('heading', { name: /CV/i })).toBeInTheDocument()
        expect(screen.getByText('what I have done')).toBeInTheDocument()
        expect(screen.getByText('Loading...')).toBeInTheDocument()

        expect(startSpy).toHaveBeenCalledTimes(0)

        expect(container).toMatchSnapshot()
    })

    it('should match snapshot when status === "done"', () => {
        const startSpy = jest.spyOn(experienceSlice.actions, 'initializeStart')
        const { container } = render(
            <Provider
                store={mockStore(
                    mockState({
                        experience: {
                            status: 'done',
                            jobs: {
                                [jobsArray[0].id]: { ...jobsArray[0], duration: { months: 4 } },
                                [jobsArray[1].id]: { ...jobsArray[1], duration: { months: 3 } },
                            },
                            techs: {
                                [techs.CSS]: { years: 1, months: 4 },
                                [techs.HTML]: { years: 1, months: 3 },
                            },
                            totalExperience: { years: 1, months: 7 },
                        },
                    })
                )}
            >
                <CV />
            </Provider>
        )

        expect(screen.getByRole('heading', { name: /CV/i })).toBeInTheDocument()
        expect(screen.getByText('what I have done')).toBeInTheDocument()

        expect(screen.getByRole('heading', { name: /^Technologies/i })).toBeInTheDocument()
        expect(screen.getByText(`${techs.CSS} 1 4 1 7`)).toBeInTheDocument()
        expect(screen.getByText(`${techs.HTML} 1 3 1 7`)).toBeInTheDocument()

        expect(screen.getByRole('heading', { name: /Jobs/i })).toBeInTheDocument()
        expect(screen.getByText(jobsArray[0].id)).toBeInTheDocument()
        expect(screen.getByText(jobsArray[1].id)).toBeInTheDocument()

        expect(screen.getByRole('heading', { name: /Other technologies/i })).toBeInTheDocument()

        expect(startSpy).toHaveBeenCalledTimes(0)

        expect(container).toMatchSnapshot()
    })
})
