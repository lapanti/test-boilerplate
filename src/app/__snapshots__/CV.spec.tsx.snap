// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CV /> should match snapshot when status === "" 1`] = `
.c0 {
  grid-area: header;
  color: var(--accent-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: var(--l) 0;
}

.c1 {
  margin: 0;
}

.c2 {
  grid-area: main;
  display: grid;
  grid-template-columns: 1fr min(65ch,calc(100% - var(--xxl))) 1fr;
  grid-column-gap: var(--xl);
  grid-auto-rows: minmax(min-content,max-content);
}

.c2 > * {
  grid-column: 2;
}

@media (min-width:660px) {
  .c2 {
    grid-template-columns: 1fr min(65ch,100%) 1fr;
    grid-column-gap: normal;
  }
}

<div>
  <header
    class="c0"
  >
    <h1
      class="c1"
    >
      CV
    </h1>
    <p>
      what I have done
    </p>
  </header>
  <main
    class="c2"
  >
    <p>
      I started out as most people, doing websites in my early teens (I'll admit, my first ones were with FrontPage) for my bands etc. I once even borrowed (and did all the exercises) from the library a book that detailed the then popular HTML, PHP, MySQL stack, so a lot of the websites I did back then used PHP on the level of 
      <code>
        include
      </code>
      . Luckily I outgrew that and started dabbling in jQuery and finally got a formal education as well.
    </p>
    <p>
      That all lead into a (tech) master's degree in Information Networks from Aalto University, where I majored in ICT in business (basically all the software courses I could find). Apart from that I have accumulated some professional experience in the field
      
      .
    </p>
    <p>
      Loading...
    </p>
  </main>
</div>
`;

exports[`<CV /> should match snapshot when status === "done" 1`] = `
.c0 {
  grid-area: header;
  color: var(--accent-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: var(--l) 0;
}

.c1 {
  margin: 0;
}

.c2 {
  grid-area: main;
  display: grid;
  grid-template-columns: 1fr min(65ch,calc(100% - var(--xxl))) 1fr;
  grid-column-gap: var(--xl);
  grid-auto-rows: minmax(min-content,max-content);
}

.c2 > * {
  grid-column: 2;
}

.c3 {
  list-style-type: none;
  padding: 0;
}

@media (min-width:660px) {
  .c2 {
    grid-template-columns: 1fr min(65ch,100%) 1fr;
    grid-column-gap: normal;
  }
}

<div>
  <header
    class="c0"
  >
    <h1
      class="c1"
    >
      CV
    </h1>
    <p>
      what I have done
    </p>
  </header>
  <main
    class="c2"
  >
    <p>
      I started out as most people, doing websites in my early teens (I'll admit, my first ones were with FrontPage) for my bands etc. I once even borrowed (and did all the exercises) from the library a book that detailed the then popular HTML, PHP, MySQL stack, so a lot of the websites I did back then used PHP on the level of 
      <code>
        include
      </code>
      . Luckily I outgrew that and started dabbling in jQuery and finally got a formal education as well.
    </p>
    <p>
      That all lead into a (tech) master's degree in Information Networks from Aalto University, where I majored in ICT in business (basically all the software courses I could find). Apart from that I have accumulated some professional experience in the field
       (1 years and 7 months, to be exactish)
      .
    </p>
    <h2>
      Technologies
    </h2>
    <ul
      class="c3"
    >
      <span>
        CSS
         1 4
         1 7
      </span>
      <span>
        HTML
         1 3
         1 7
      </span>
    </ul>
    <h2>
      Jobs
    </h2>
    <span>
      pulmatonSolutionsOy
    </span>
    <span>
      nokiaSolutionsAndNetworks
    </span>
    <h2>
      Other technologies
    </h2>
    <p>
      I have also used or dabbled with some other technologies, listed below, but I do not consider myself fluent (or willing) to work with them professionally:
       
    </p>
    <ul
      class="c3"
    >
      <li>
        Docker
      </li>
      <li>
        Java
      </li>
      <li>
        Python
      </li>
      <li>
        PHP
      </li>
      <li>
        MySQL
      </li>
      <li>
        PostgreSQL
      </li>
      <li>
        React Native
      </li>
      <li>
        Android
      </li>
      <li>
        And the list goes on and on
      </li>
    </ul>
  </main>
</div>
`;

exports[`<CV /> should match snapshot when status === "loading" 1`] = `
.c0 {
  grid-area: header;
  color: var(--accent-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: var(--l) 0;
}

.c1 {
  margin: 0;
}

.c2 {
  grid-area: main;
  display: grid;
  grid-template-columns: 1fr min(65ch,calc(100% - var(--xxl))) 1fr;
  grid-column-gap: var(--xl);
  grid-auto-rows: minmax(min-content,max-content);
}

.c2 > * {
  grid-column: 2;
}

@media (min-width:660px) {
  .c2 {
    grid-template-columns: 1fr min(65ch,100%) 1fr;
    grid-column-gap: normal;
  }
}

<div>
  <header
    class="c0"
  >
    <h1
      class="c1"
    >
      CV
    </h1>
    <p>
      what I have done
    </p>
  </header>
  <main
    class="c2"
  >
    <p>
      I started out as most people, doing websites in my early teens (I'll admit, my first ones were with FrontPage) for my bands etc. I once even borrowed (and did all the exercises) from the library a book that detailed the then popular HTML, PHP, MySQL stack, so a lot of the websites I did back then used PHP on the level of 
      <code>
        include
      </code>
      . Luckily I outgrew that and started dabbling in jQuery and finally got a formal education as well.
    </p>
    <p>
      That all lead into a (tech) master's degree in Information Networks from Aalto University, where I majored in ICT in business (basically all the software courses I could find). Apart from that I have accumulated some professional experience in the field
      
      .
    </p>
    <p>
      Loading...
    </p>
  </main>
</div>
`;
